### PR TITLE
Space - Shonda & Diana

### PR DESCRIPTION
**Pull request by dnguye2**
# Assignment Submission: OO Ride Share

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? |  The first day, it was very overwhelming to understand what was going on. Using pry helped to understand how the codebase worked. Next time, it would be helpful to look at every file and also examine the tests. This would have expedited understanding of the codebase. |
What _inheritance_ relations exist between classes? | Within the trip dispatcher, which is the holder, csv_record was the super class of the following sub classes which had inheritance from csv_record: drivers, passengers, and trips. |
What _composition_ relations exist between classes? | The driver class and the passenger class has a trip. |
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |  For the net expenditures method, we had to handle when the @trips array was nil. We were trying to decide to use a `nil?` which didn't work at first. Then we realized we had use `.empty?` because `nil?` refers nothing whereas @trips had an empty array which is not nil. This led us to realize `.empty?` was the best choice because it covered the case we were looking for. |
Give an example of a _template method_ that you implemented for this assignment | For the driver class, the template method we created and used was `def self.from_csv(record)`.|
Give an example of a _nominal test_ that you wrote for this assignment |  For the passenger `net_expenditures` method, we wrote a nominal test for `it "returns the correct total amount of money that a passenger has spent"`|
Give an example of an _edge case test_ that you wrote for this assignment | For the driver `average_rating` we wrote an edge case test for `it "returns zero if no driven trips" do` and `it "returns 0 if trip costs <= $1.65" do` |
What is a concept that you gained more clarity on as you worked on this assignment |  We got more clarity on how classes function in relation to each other. And as a whole, object oriented programming and design. In addition, we gained more clarity on handling nil cases. |

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  
+  def initialize(id:, name:, vin:, status: :AVAILABLE, trips: nil)
+    super(id)
+
+    @name = name
+    @vin = vin
+    @status = status.to_sym
+    @trips = trips || []
+  
+    vin_pattern = /^[A-Z0-9]{17}$/i
```
Comment (kaidamasaki) [inline lib/driver.rb:16]: Regular expressions!  :tada:

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    if (vin_pattern =~ @vin) == nil
+      raise ArgumentError, "Invalid VIN"
```
Comment (kaidamasaki) [inline lib/driver.rb:18]: It's helpful to include the bad argument in the message for your `ArgumentError`s:

```suggestion
      raise ArgumentError, "Invalid VIN: #{@vin}"
```

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    end
+  end
+
+  def add_trip(trip)
+    @trips << trip
+  end
+
+  def change_status
```
Comment (kaidamasaki) [inline lib/driver.rb:26]: The name `change_status` is misleading since it just sets the status to `:UNAVAILABLE`.

```suggestion
  def make_unavailable
```

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          nil_cases += 1
+        end
+      end
+
+      average_rating = (@trips).map do |trip|
+        if trip.rating == nil
+          0
+        else
+        (trip.rating).to_f
+        end
```
Comment (kaidamasaki) [inline lib/driver.rb:48]: Incorrect indentation:

```suggestion
        else
          (trip.rating).to_f
        end
```

File: lib/driver.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      end
+
+      
+      cost = ((average_rating).inject(:+))/((average_rating.length) - nil_cases)
+      return cost
```
Comment (kaidamasaki) [inline lib/driver.rb:53]: You have a lot of extra parenthesis and can directly return the result of this computation, also `inject(:+)` is the same as `sum`):

```suggestion
      return average_rating.sum / (average_rating.length - nil_cases)
```

Remember, `.` has higher precedence than arithmetic.

File: lib/driver.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    end
+  end
```
Comment (kaidamasaki) [inline lib/driver.rb:55]: You can simplify this method using `reject` and `sum`.

```suggestion
  def average_rating
    return 0 if trips.empty?

    rated_trips = trips.reject { |trip| trip.rating.nil? }
    
    return rated_trips.sum { |trip| trip.rating } / rated_trips.length
  end
```

File: lib/driver.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  def change_status
+    @status = :UNAVAILABLE
+  end
+
+  # Wave 2 get average rating for a driver
+  def average_rating
+    nil_cases = 0
+
+    if @trips.empty? == true
```
Comment (kaidamasaki) [inline lib/driver.rb:34]: You don't need to compare boolean values to `true`:

```suggestion
    if @trips.empty?
```

File: lib/trip.rb [hunk 8]
```diff
@@ -1,41 +1,68 @@
...
       end
+
+      # 1.1 ArgumentError for end time < start time
+      if @end_time.nil? == true
+        @end_time = @end_time
+      elsif @start_time.nil? == true
+        @start_time = @start_time
+      elsif @end_time >= @start_time
+        @end_time = @end_time
+        @start_time = @start_time
```
Comment (kaidamasaki) [inline lib/trip.rb:62]: I'm not clear what you're doing here.  

In all of these conditions you're just setting `@start_time` to itself or `@end_time` to itself which doesn't have any effect.

File: test/passenger_test.rb [hunk 9]
```diff
@@ -68,7 +73,79 @@
...
+          driver: @driver,
+          start_time: "#{Time.now - 60 * 60}",
+          end_time: "#{Time.now + 60 * 60}",
+          rating: 5,
+          cost: 10
+          )
+        @passenger.add_trip(trip)
+      end
+      
+      expect (_@passenger.net_expenditures).must_equal 20
```
Comment (kaidamasaki) [inline test/passenger_test.rb:106]: Using the `_` with `expect` is unnecessary:

```suggestion
      expect(@passenger.net_expenditures).must_equal 20
```

File: test/passenger_test.rb [hunk 10]
```diff
@@ -68,7 +73,79 @@
...
+
+  # Tests for 1.2 total_time_spent
+  describe "total_time_spent" do
+    before do
+      @passenger = RideShare::Passenger.new(
+        id: 9,
+        name: "Merl Glover III",
+        phone_number: "1-602-620-2330 x3723",
+        trips: []
+        )
```
Comment (kaidamasaki) [inline test/passenger_test.rb:122]: Incorrect indentation:

```suggestion
      @passenger = RideShare::Passenger.new(
        id: 9,
        name: "Merl Glover III",
        phone_number: "1-602-620-2330 x3723",
        trips: []
      )
```

File: lib/trip.rb [hunk 11]
```diff
@@ -46,19 +73,28 @@ def inspect
...
+    def connect(passenger, driver)
       @passenger = passenger
+      @driver = driver
       passenger.add_trip(self)
+      driver.add_trip(self)
+    end
+
+    # 1.1 #4 instance method of duration of trip
+    def duration
+      return (Time.parse(@end_time) - Time.parse(@start_time)).to_i
```
Comment (kaidamasaki) [inline lib/trip.rb:85]: You're already parsing the time in `from_csv` so this is redundant.

File: test/passenger_test.rb [hunk 12]
```diff
@@ -68,7 +73,79 @@
...
+    end
+
+    it "returns the total time that passenger has spent on their trips" do
+      2.times do
+        trip = RideShare::Trip.new(
+          id: 8,
+          passenger: @passenger,
+          driver: @driver,
+          start_time: "#{Time.now - 60 * 60}",
+          end_time: "#{Time.now + 60 * 60}",
```
Comment (kaidamasaki) [inline test/passenger_test.rb:137]: You should just pass `Time` objects into the test.

```suggestion
          start_time: Time.now - 60 * 60,
          end_time: Time.now + 60 * 60,
```

File: lib/trip_dispatcher.rb [hunk 13]
```diff
@@ -27,14 +37,54 @@ def inspect
...
+      return @available_driver
+    end
+
+    # Wave 3 method that allows for an in-progress trip to be made
+    def request_trip(passenger_id)
+      @passenger = self.find_passenger(passenger_id)
+     
+      # handles case when there are no available drivers
+      if @drivers.find {|driver| driver.status == :AVAILABLE}.nil?
+        raise ArgumentError.new("No available drivers.")
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb:52]: This isn't an issue with the arguments here.  You should use a custom error class here.

For example at the top of the file you can define:
```ruby
  class NoDriverError < StandardError
  end
```

And then: 
```suggestion
        raise NoDriverError.new("No available drivers.")
```

File: lib/trip_dispatcher.rb [hunk 14]
```diff
@@ -27,14 +37,54 @@ def inspect
...
+      end
+
+      # grabs first available driver
+      @driver = select_available_driver
+      @driver.change_status
+
+      @trip_data = {
+        id: @trips.last.id + 1,
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb:60]: Nice use of `last` here to assign a unique id.

**Comment by kaidamasaki** [comment]
# OO Ride Share



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The code demonstrates individual learning about `Time` and the responsibility of `Trip.from_csv`, and uses `Time.parse` in `Trip.from_csv` | Somewhat.
The code demonstrates breaking out complex logic in helper methods, such as making a helper method in `Trip` to calculate duration | ✔️
There are tests for the nominal cases for the `Passenger#net_expenditures` and `Passenger#total_time_spent` | ✔️
There is at least one edge case test for either `Passenger#net_expenditures` or `Passenger#total_time_spent` testing if the passenger has no trips | ✔️
Practices inheritance. `Driver` inherits from `CsvRecord`, and implements `from_csv` | ✔️
Employs problem-solving and implements `Driver#average_rating` and `Driver#total_revenue` | ✔️
Implements the `TripDispatcher#request_trip`, which creates an instance of `Trip` with a driver and passenger, adds the new trip to `@trips`, and changes the status of the driver | ✔️
Practices composition. In `TripDispatcher#request_trip`, the driver gets connected to the new trip, the passenger gets connected to the new trip | ✔️
Practices git with at least 10 small commits and meaningful commit messages | ✔️

## Testing Requirements

| Testing Requirement | yes/no |
| --- | --- |
There is reasonable test coverage for wave 1, and all wave 1 tests pass | ✔️
There is reasonable test coverage for wave 2, and all wave 2 tests pass | ✔️
Wave 3: Tests in wave 1 and wave 2 explicitly test that only completed trips should be calculated (and ignore in-progress trips) |
There is reasonable test coverage for `TripDispatcher#request_trip`, and all tests pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 3+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 6+ in Code Review && 2+ in Functional Requirements |
| Red (Not at Standard) | 0-5 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | (See comments.)
| Elegant/Clever |
| Descriptive/Readable | ✅
| Concise |
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Good job!  Here are some tips for how you can clean up your code. 😄
